[zechary28] iP

zechary28 opened the pull request:
### Pull Request: *Luke the Chatbot Task Manager* 🚀  

**Description:**  
This PR introduces *Luke*, a chatbot task manager that's here to save your productivity galaxy. Equipped with Jedi-like focus, Luke will help you manage tasks, update priorities, and remind you that *procrastination leads to the dark side.*

---

#### Features:  
- Luke keeps track of your tasks with **Jedi discipline** and *sassy remarks*.  
- Offers **quick updates** using `inline code` for precision.  
- Sends reminders with quotes like, *“Do or do not. There is no try.”*  
- Provides **humor mode** because *managing tasks doesn't have to be boring*.  
- Lies about its features

---

#### How to Use:  
1. Add Luke to your task management system.  
2. Assign tasks with:  
   ```
   todo lightsaber training
   deadline destroy death star /by 01/01/3999 23:59
   event fight dad /from 02/02/4000 12:00 /to 03/02/4000 12:00
   ```
3. Mark tasks as complete with:  
   ```
   mark 1
   ```  

---

#### Remaining Tasks:  
- [x] MVP.  
- [ ] Provide user feedback for exceptions.  
- [ ] Add character.  
- [ ] Create file in dedicated directory for .jar  

---

> *"This is the task manager you're looking for."*  

For more information, check out [Luke's Galactic Documentation](https://github.com/zechary28/ip).  

---
Here is some `inline code`.  

**May your tasks be few, and your deadlines met.**  
**~ The Dev Team**

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

zechary28 committed:
added gitignore and changed name to Luke

zechary28 committed:
added question and bye

zechary28 committed:
added printLine and exit functions

zechary28 committed:
added printLine and exit functions

zechary28 committed:
added printList function

zechary28 committed:
add input into array

zechary28 committed:
fixed list adding

zechary28 committed:
make Task class

zechary28 committed:
added Task.toString

zechary28 committed:
added mark unmark functionality

zechary28 committed:
created Deadline class, changed Task class to abstract

zechary28 committed:
created ToDo class, amended Deadline.toString

zechary28 committed:
made Event class and fixed Deadline and ToDo

zechary28 committed:
input parsing and functionality for task subtypes

zechary28 committed:
Text UI test

zechary28 committed:
refactor handle Task types

zechary28 committed:
error handling for todo and deadline invalid input

zechary28 committed:
finished invalid input error handling for task types

zechary28 committed:
handle NumberFormatException

zechary28 committed:
added delete functionality and text-ui-test for delete and invalid input

zechary28 committed:
Create list.txt file and create list reading functions

zechary28 committed:
Handle errors when file found or not, wrap main loop in try catch statement

zechary28 committed:
Enable read and write from task list file

zechary28 committed:
Fix list file

zechary28 committed:
Merge branch 'branch-Level-7'

zechary28 committed:
Change Task time field from String to LocalDateTime, Task constructors takes in formatted string

zechary28 committed:
Format date from DD/MM/YYYY HH:MM input to YYYY:MM:DDTHH:MM:SS and vice versa

zechary28 committed:
Add list file and backup for read and write

zechary28 committed:
Merge branch 'branch-Level-8'

zechary28 committed:
Create Command, TaskList, Ui, Storage classes with elementary methods

zechary28 committed:
Change functions to access tasklist and file via Tasklist and Storage instances, create Luke.run(), not done integrating with main

zechary28 committed:
Move main logic into run, run logic calls changed functions from prev commit

zechary28 committed:
Fix bugs and tested, working

zechary28 committed:
Create packages duke.ui, duke.component, duke.excpetions, duke.command, duke.task, Move code into respective packages

zechary28 committed:
Merge branch 'add-gradle-support' of https://github.com/zechary28/ip

zechary28 committed:
Enable build and run for Gradle, put Task subtypes in their own files

WeiHungLoh commented on src/main/java/duke/ui/Luke.java:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    public Task readTask(String input) throws InvalidInputException {
+        String[] task = input.split(" : ");
```
Perhaps String[] tasks may be more suitable here since the array consists of task type, description and deadlines?

WeiHungLoh commented on src/main/java/duke/ui/Luke.java:
```diff
@@ -0,0 +1,273 @@
+                else if (command.equals("list")) printList();
+                else if (command.equals("delete")) deleteTask(Integer.parseInt(inputArr[1]) - 1);
+                else { // add tasks
+                    String type = command;
```
Perhaps declaring this as String commandType = command to be more specific would be better?

WeiHungLoh commented on src/main/java/duke/component/Storage.java:
```diff
@@ -0,0 +1,51 @@
+
+public class Storage {
+    private final String READ_FILE_PATH = "../../../data/list.txt";
+    private final String WRITE_FILE_PATH = "../../../data/list.txt";
```
Perhaps you can consider grouping them under private static final String FILE_PATH = "../../../data/list.txt" since they have the same file path?

WeiHungLoh commented on src/main/java/duke/ui/Luke.java:
```diff
@@ -0,0 +1,273 @@
+        if (start.trim().isEmpty() || end.trim().isEmpty()) throw new InvalidInputException();
+        return new Event(name, false, start, end);
+    }
+
```
I like how you create separate parse methods for the respective tasks rather than group them together under parseTasks(...).

WeiHungLoh reviewed (commented):
Looks good to me overall in terms of adhering to naming guidelines. Keep up the good work!

zechary28 committed:
Create stubs for task subtypes

zechary28 committed:
Create tests for TaskListTest, passed 5/5

zechary28 committed:
Add JavaDoc for packages luke.component, luke.task, luke.ui

zechary28 committed:
Add findtask(String) but its O(n), implement in Luke.run()

zechary28 committed:
Merge branch 'branch-A-JavaDoc'

zechary28 committed:
Tweak coding standard

zechary28 committed:
Merge master with branch-Level-9

WhiteBear82 commented on src/main/java/luke/component/Storage.java:
```diff
@@ -0,0 +1,96 @@
+ */
+public class Storage {
+    private final String READ_FILE_PATH = "data/list.txt";
+    private final String WRITE_FILE_PATH = "data/list.txt";
```
Using all caps and underscores for constant variables is a good practice. It helps improve readability and clearly distinguishes constants from other variables. Nice work! 👍

WhiteBear82 commented on src/test/java/luke/component/StorageTest.java:
```diff
@@ -0,0 +1,55 @@
+public class StorageTest {
+    private final String READ_FILE_PATH = "data/list.txt";
+    private final String WRITE_FILE_PATH = "data/list.txt";
+    Scanner scanner;
```
The field name follows the naming convention, but perhaps making it a bit more specific, like fileScanner, could help distinguish it from other potential uses of scanners in the future. It might improve clarity and make the code easier to understand.

WhiteBear82 commented on src/main/java/luke/component/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+ */
+public class TaskList {
+
+    private ArrayList<Task> list;
```
The name is a bit generic, so using something more specific like `taskList` might help make its purpose clearer and improve readability.

WhiteBear82 commented on src/test/java/luke/exception/InvalidInputExceptionTest.java:
```diff
@@ -0,0 +1,4 @@
+package luke.exception;
+
+// exceptions
+public class InvalidInputExceptionTest extends Exception {}
```
The class name follows conventions well and clearly reflects its purpose. 👍

WhiteBear82 commented on src/main/java/luke/task/Deadline.java:
```diff
@@ -0,0 +1,81 @@
+     *
+     * @return {@code true} if the task is done, {@code false} otherwise
+     */
+    public boolean getIsDone() {
```
I'm not sure what convention our course follows for getters, but the name is functional. It might be worth considering simplifying it to isDone to better align with Java naming conventions for boolean getters, if that works for the style we're aiming for.

WhiteBear82 commented on src/main/java/luke/ui/Luke.java:
```diff
@@ -0,0 +1,393 @@
+            throw new InvalidInputException();
+        }
+        String name = inputArr[0];
+        String due = inputArr[1];
```
It might be helpful to replace "due" with something more descriptive, like dueDate, as "due" could refer to different things depending on the context. For example, "due" could refer to a payment due date, which might be confusing in this context. Using a more specific name like dueDate would make the purpose clearer. Using a more specific name would improve clarity.

WhiteBear82 reviewed (commented):
I really admire how consistent you’ve been with following the course conventions, especially with using verbs for methods and nouns for classes. It really helps make the code clear and easy to understand. Keep up the good work!

shashwatchan commented on src/main/java/luke/ui/Luke.java:
```diff
@@ -0,0 +1,393 @@
+                else if (command.equals("unmark")) markTask(Integer.parseInt(inputArr[1]) - 1, false);
+                else if (command.equals("list")) printList();
+                else if (command.equals("delete")) deleteTask(Integer.parseInt(inputArr[1]) - 1);
+                else if (command.equals("find")) findTask(inputArr[1]);
```
I think the coding standard mentions that conditionals should all be in braces and on separate lines. 
Also, perhaps it might be cleaner to use a switch statement for parsing the command here?

shashwatchan commented on src/main/java/luke/ui/Luke.java:
```diff
@@ -0,0 +1,393 @@
+    public Luke() {
+        this.taskList = new TaskList();
+        this.ui = new Ui();
+        try {
```
Perhaps add a finally block here as well to comply with the coding standard?

shashwatchan commented on src/main/java/luke/ui/Luke.java:
```diff
@@ -0,0 +1,393 @@
+
+import luke.task.*;
+import luke.component.*;
+import luke.exception.*;
```
Perhaps you can explicitly list the classes (especially for those with less than 3 imports) here to comply with the coding standard?

shashwatchan commented on src/test/java/luke/ui/LukeTest.java:
```diff
@@ -0,0 +1,279 @@
+            writeListToFile();
+        } catch (Exception e) {
+
+        }
```
Perhaps handle this exception so the program does not fail silently?

shashwatchan reviewed (commented):
I think the PR overall is pretty well done. I've just added some minor comments relating to the coding standard.

zechary28 committed:
Add Checkstyle, fix style errors for main

zechary28 committed:
Add JavaFX

zechary28 committed:
Refactor input and output for JavaFX, working

zechary28 committed:
Merge branch 'branch-Level-10'

zechary28 committed:
Move list file to main/resources/data, change read and write paths

zechary28 committed:
Refactor functions in Luke.java

zechary28 committed:
Add assertions to task parsing

zechary28 committed:
Merge pull request #2 from zechary28/branch-A-Assertions

Add assertions to task parsing

zechary28 committed:
Merge with updated master branch

zechary28 committed:
Merge pull request #3 from zechary28/branch-A-CodeQuality

Branch a code quality

zechary28 committed:
Add sort feature with Task.compareTo()

zechary28 committed:
Enable read and write for JAR

Storage reads and writes to /src/main/resources/data/list.txt,
only works in development environment.

Running JAR file does not have paths available,
will fail to read and write.

Enable read and write features to JAR.
For reading, Storage class will check which read path to use
to see if program is running in environment or JAR.
For writing, Storage will create file and directory if doesn't exist
and write to respective file in respective path.

zechary28 committed:
Handle array out of bounds for mark and delete

zechary28 committed:
Enable case-insensitive commands

zechary28 committed:
Change ui font to monospace

zechary28 committed:
Improve error handling and user friendliness

There is no guidance and feedback for users.
Usually shows "I don't understand" for most exceptions.

Not very user friendly because commands are not specifies,
and very specfic, not case sensitive, unknown format.

Add help command to display list of all available commands.
Add feedback to user when wrong format is given for tasks.
Enable case-insensitive commands.

zechary28 committed:
Add Javadoc comments, clean up code style

zechary28 committed:
Add user guide in README.md

zechary28 committed:
Add credit for background image

zechary28 committed:
Edit readme.md
